spinta/commands/write.py

```diff
@@ -864,6 +865,68 @@ async def prepare_data_for_write(
         yield data
 
 
+async def sort_for_foreign_key_dependency(
+    dstream: AsyncIterator[DataItem],
+    action,
+    model: Model
+) -> AsyncIterator[DataItem]:
+    # Get properties from model, whose ref is self
+    self_reference_properties = [prop.name for prop in model.properties.values() if
+                                 isinstance(prop.dtype, Ref) and prop.dtype.model == model]
+    data_with_self_reference = {}
+
+    # Get self ref properties from payload and check if they are all None
+    # if all None means no need to check for foreign key dependency and just Yield
+    # if action is not INSERT or UPSERT, it should skip sorting, since DELETE can be dependent on data that is not in stream
+    # else store the data for further processing
+    async for data in dstream:
+        payload = data.payload
+        filtered_properties = [payload.get(key) for key in self_reference_properties]
+        if all(value is None for value in filtered_properties) or action not in [Action.INSERT, Action.UPSERT]:
+            yield data
+        else:
+            data_with_self_reference[payload.get("_id")] = {
+                "data": data,
+                "self_reference": filtered_properties
+            }
```

Comment:
Unfortunately this won't work with large datasets, because as I understand, here all table data is loaded into memory and then sorted?

But of course this would be a good solution if all datasets would fit into memory.

If data would be sorted using SQL query this would work. But in order to sort data using SQL we need a recursive CTE, see https://stackoverflow.com/questions/59137122/sql-query-for-topological-sort